beetsplug/mosaic.py

```diff
@@ -0,0 +1,249 @@
+
+import os.path
+
+from beets.plugins import BeetsPlugin
+from beets import ui
+
+from PIL import Image
+import math
+from parse import *
+
+# MOSAICFONT = os.path.join(os.path.dirname(__file__), 'FreeSans.ttf')
+
+
+class MosaicCoverArtPlugin(BeetsPlugin):
+    col_size = 4
+
+    def __init__(self):
+        super(MosaicCoverArtPlugin, self).__init__()
+
+        self.config.add({'geometry': '100x100+3+3',
+                         'tile': '',  # or e.g. x3 3x
+                         'label': '',
+                         'background': 'ffffff',
+                         'mosaic': 'mosaic.png',
+                         'show_mosaic': False,
+                         'watermark': '',
+                         'watermark_alpha': 0.4})
+
+    def commands(self):
+        cmd = ui.Subcommand('mosaic', help=u"create mosaic from coverart")
+
+        cmd.parser.add_option(
+            u'-m', u'--mosaic', dest='mosaic',  metavar='FILE',
+            action='store',
+            help=u'save final mosaic picture as FILE'
+        )
+
+        cmd.parser.add_option(
+            u'-w', u'--watermark', dest='watermark',
+            action='store', metavar='FILE',
+            help=u'add FILE for a picture to blend over mosaic'
+        )
+
+        cmd.parser.add_option(
+            u'-a', u'--alpha', dest='watermark_alpha',
+            action='store', metavar='ALPHA',
+            help=u'ALPHA value for blending'
+        )
+        cmd.parser.add_option(
+            u'-c', u'--color', dest='background',
+            action='store', metavar='HEXCOLOR',
+            help=u'background color as HEXCOLOR'
+        )
+
+        cmd.parser.add_option(
+            u'-g', u'--geometry', dest='geometry',
+            action='store', metavar='GEOMETRY',
+            help=u'define geometry as <width>x<height>+<marginx>+<marginy>'
+        )
+
+        def func(lib, opts, args):
+            self.config.set_args(opts)
+
+            if self.config['mosaic']:
+                mosaic = opts.mosaic or self.config['mosaic'].get(str)
+            else:
+                mosaic = opts.mosaic
+
+            self._log.debug(u'Mosaic: {}', mosaic)
+
+            if self.config['watermark']:
+                watermark = opts.watermark or self.config['watermark'].get(str)
+            else:
+                watermark = opts.watermark
+
+            watermark_alpha = float(opts.watermark_alpha) or self.config[
+                'watermark_alpha'].get(float)
+
+            if self.config['background']:
+                background = opts.background or self.config[
+                    'background'].get(str)
+            else:
+                background = opts.background
+
+            if self.config['geometry']:
+                geometry = opts.geometry or self.config[
+                    'geometry'].get(str)
+            else:
+                geometry = opts.background
+            albums = lib.albums(ui.decargs(args))
```

Comment:
Of course, you don't even need to pass around stuff like `geometry` as a parameter—the code that needs it can just use `self.config['geometry'].as_str()` right there.